### ISSUE TITLE
Redraw bauhaus widgets before updating preview

### ISSUE THREAD

**Issue by aurelienpierre** [aurelienpierreeng/ansel#29]
Changing bauhaus slider values seems to trigger a pipeline recompute first, then a GUI redraw. For heavy edits, this introduces a noticeable lag between user action and GUI feedback.

GUI elements need to be redrawn before recomputing pipe. But it is possible that the lag is also caused by thread mutex locks.

To investigate.

**Comment by aurelienpierre** [comment]
Investigation lead to concerning findings. For bauhaus sliders, the new parameter value is set on `button_pressed` and then again on `button_released` events. Each of these settings triggers a new pipeline recompute, which is a massive loss of performance. The first fix is to set values only on `button_released` but that doesn't improve the overall responsiveness much.

**Comment by aurelienpierre** [comment]
I have been at it for more than 10 h now, and the amount of technical debt accumulated since 2021 on this code is outstanding. 

Lots of things are subtly broken, like the slider coordinates are sometimes computed taking GUI margins/padding into account, sometimes not, because nobody reused the getter function I wrote in 2019 to compute them all the same all the time, but copy-pasted the same line of code that someone else did not update everywhere later.

The "curve" used in slider is applied twice in the typical slider, but the support is not implemented in the magnified controller in pop-up. Since it's used only in ashift.c for paramters typically computed automatically by the solver, it's a silent bug, so I removed that feature entirely as there is no reason to complexify the code this way.

Functions have been hacked without refactoring, such that we recompute the same widths and heights many times instead of storing them and end-up with function taking 14 parameters, like `show_pango_text(w, context, cr, max, w2 - _widget_get_quad_width(w), ht + INNER_PADDING, 0, TRUE, FALSE, PANGO_ELLIPSIZE_END, FALSE, FALSE, NULL, NULL);`. Good luck untangling that.

Branching has been nested and is non-uniform, such that the only way to figure out the sequence of execution is to `printf(stdout)` in each function and to look up at the console output. Enjoy the glorious unlegibility of 

```C
static float _action_process_slider(gpointer target, dt_action_element_t element, dt_action_effect_t effect,
                                    float move_size)
{
  GtkWidget *widget = GTK_WIDGET(target);
  dt_bauhaus_widget_t *bhw = DT_BAUHAUS_WIDGET(widget);
  dt_bauhaus_slider_data_t *d = &bhw->data.slider;

  if(!isnan(move_size))
  {
    switch(element)
    {
      case DT_ACTION_ELEMENT_VALUE:
      case DT_ACTION_ELEMENT_FORCE:
        switch(effect)
        {
          case DT_ACTION_EFFECT_POPUP:
            dt_bauhaus_show_popup(widget);
            break;
          case DT_ACTION_EFFECT_DOWN:
            move_size *= -1;
          case DT_ACTION_EFFECT_UP:
            ++d->is_dragging;
            _slider_add_step(widget, move_size, GDK_MODIFIER_MASK, element == DT_ACTION_ELEMENT_FORCE);
            --d->is_dragging;
            break;
          case DT_ACTION_EFFECT_RESET:
            dt_bauhaus_slider_reset(widget);
            break;
          case DT_ACTION_EFFECT_TOP:
            dt_bauhaus_slider_set_from_param(widget, element == DT_ACTION_ELEMENT_FORCE ? d->hard_max : d->max);
            break;
          case DT_ACTION_EFFECT_BOTTOM:
            dt_bauhaus_slider_set_from_param(widget, element == DT_ACTION_ELEMENT_FORCE ? d->hard_min : d->min);
            break;
          case DT_ACTION_EFFECT_SET:
            dt_bauhaus_slider_set_from_param(widget, move_size);
            break;
          default:
            fprintf(stderr, "[_action_process_slider] unknown shortcut effect (%d) for slider\n", effect);
            break;
        }

        gchar *text = dt_bauhaus_slider_get_text(widget, dt_bauhaus_slider_get(widget));
        dt_action_widget_toast(bhw->module, widget, text);
        g_free(text);

        break;
      case DT_ACTION_ELEMENT_BUTTON:
        _action_process_button(widget, effect);
        break;
      case DT_ACTION_ELEMENT_ZOOM:
        switch(effect)
        {
          case DT_ACTION_EFFECT_POPUP:
            dt_bauhaus_show_popup(widget);
            break;
          case DT_ACTION_EFFECT_RESET:
            move_size = 0;
          case DT_ACTION_EFFECT_DOWN:
            move_size *= -1;
          case DT_ACTION_EFFECT_UP:
            _slider_zoom_range(bhw, move_size);
            break;
          case DT_ACTION_EFFECT_TOP:
          case DT_ACTION_EFFECT_BOTTOM:
            if((effect == DT_ACTION_EFFECT_TOP) ^ (d->factor < 0))
              d->max = d->hard_max;
            else
              d->min = d->hard_min;
            gtk_widget_queue_draw(widget);
            break;
          default:
            fprintf(stderr, "[_action_process_slider] unknown shortcut effect (%d) for slider\n", effect);
            break;
        }

        _slider_zoom_toast(bhw);
        break;
      default:
        fprintf(stderr, "[_action_process_slider] unknown shortcut element (%d) for slider\n", element);
        break;
    }
  }

  if(element == DT_ACTION_ELEMENT_BUTTON) return dt_bauhaus_widget_get_quad_active(widget);

  if(effect == DT_ACTION_EFFECT_SET) return dt_bauhaus_slider_get(widget);

  return d->pos
         + (d->min == -d->max ? DT_VALUE_PATTERN_PLUS_MINUS
                              : (d->min == 0 && (d->max == 1 || d->max == 100) ? DT_VALUE_PATTERN_PERCENTAGE : 0));
}
```

Stuff like that is simply horrendous to read :

```C
  // draw type specific content:
  cairo_save(cr);
  cairo_set_line_width(cr, 1.0);
  switch(w->type)
  {
    case DT_BAUHAUS_COMBOBOX:
    {
      // draw label and quad area at right end
      set_color(cr, *text_color);
      if(w->show_quad) dt_bauhaus_draw_quad(w, cr, inner_w, inner_h);

      dt_bauhaus_combobox_data_t *d = &w->data.combobox;
      const PangoEllipsizeMode combo_ellipsis = d->entries_ellipsis;
      const gchar *text = d->text;
      if(d->active >= 0 && d->active < d->entries->len)
      {
        const dt_bauhaus_combobox_entry_t *entry = g_ptr_array_index(d->entries, d->active);
        text = entry->label;
      }
      set_color(cr, *text_color);

      const float available_width = inner_w - _widget_get_quad_width(w);

      // calculate total widths of label and combobox
      gchar *label_text = _build_label(w);
      float label_width = 0;
      float label_height = 0;
      // we only show the label if the text is aligned on the right
      if(label_text && w->show_label && d->text_align == DT_BAUHAUS_COMBOBOX_ALIGN_RIGHT)
        show_pango_text(w, context, cr, label_text, 0, 0, 0, FALSE, TRUE, PANGO_ELLIPSIZE_END, FALSE, TRUE,
                        &label_width, &label_height);
      float combo_width = 0;
      float combo_height = 0;
      show_pango_text(w, context, cr, text, available_width, 0, 0, TRUE, TRUE, combo_ellipsis, FALSE, FALSE,
                      &combo_width, &combo_height);
      // we want to center the text verticaly
      w->top_gap = floor((inner_h - fmaxf(label_height, combo_height)) / 2.0f);
      // check if they fit
      if((label_width + combo_width) > available_width)
      {
        if(d->text_align == DT_BAUHAUS_COMBOBOX_ALIGN_RIGHT)
        {
          // they don't fit: evenly divide the available width between the two in proportion
          const float ratio = label_width / (label_width + combo_width);
          if(w->show_label)
            show_pango_text(w, context, cr, label_text, 0, w->top_gap, available_width * ratio - INNER_PADDING * 2,
                            FALSE, FALSE, PANGO_ELLIPSIZE_END, FALSE, TRUE, NULL, NULL);
          show_pango_text(w, context, cr, text, available_width, w->top_gap, available_width * (1.0f - ratio),
                          TRUE, FALSE, combo_ellipsis, FALSE, FALSE, NULL, NULL);
        }
        else if(d->text_align == DT_BAUHAUS_COMBOBOX_ALIGN_MIDDLE)
        {
          const int posx = MAX(0, (available_width - combo_width) / 2);
          show_pango_text(w, context, cr, text, posx, w->top_gap, available_width, FALSE, FALSE, combo_ellipsis,
                          FALSE, FALSE, NULL, NULL);
        }
        else
          show_pango_text(w, context, cr, text, 0, w->top_gap, available_width, FALSE, FALSE, combo_ellipsis,
                          FALSE, FALSE, NULL, NULL);
      }
      else
      {
        if(d->text_align == DT_BAUHAUS_COMBOBOX_ALIGN_RIGHT)
        {
          if(w->show_label)
            show_pango_text(w, context, cr, label_text, 0, w->top_gap, 0, FALSE, FALSE, PANGO_ELLIPSIZE_END, FALSE,
                            TRUE, NULL, NULL);
          show_pango_text(w, context, cr, text, available_width, w->top_gap, 0, TRUE, FALSE, combo_ellipsis, FALSE,
                          FALSE, NULL, NULL);
        }
        else if(d->text_align == DT_BAUHAUS_COMBOBOX_ALIGN_MIDDLE)
        {
          const int posx = MAX(0, (available_width - combo_width) / 2);
          show_pango_text(w, context, cr, text, posx, w->top_gap, 0, FALSE, FALSE, combo_ellipsis, FALSE, FALSE,
                          NULL, NULL);
        }
        else
          show_pango_text(w, context, cr, text, 0, w->top_gap, 0, FALSE, FALSE, combo_ellipsis, FALSE, FALSE, NULL,
                          NULL);
      }
      g_free(label_text);
      break;
    }
    case DT_BAUHAUS_SLIDER:
    {
      // line for orientation
      dt_bauhaus_draw_baseline(w, cr, inner_w);
      if(w->show_quad) dt_bauhaus_draw_quad(w, cr, inner_w, inner_h);

      float value_width = 0;
      if(gtk_widget_is_sensitive(widget))
      {
        cairo_save(cr);
        cairo_rectangle(cr, 0, 0, inner_w - _widget_get_quad_width(w), inner_h + INNER_PADDING);
        cairo_clip(cr);
        dt_bauhaus_draw_indicator(w, w->data.slider.pos, cr, inner_w, *fg_color, *bg_color);
        cairo_restore(cr);

        // TODO: merge that text with combo

        char *text = dt_bauhaus_slider_get_text(widget, dt_bauhaus_slider_get(widget));
        set_color(cr, *text_color);
        value_width = show_pango_text(w, context, cr, text, inner_w - _widget_get_quad_width(w), 0, 0, TRUE, FALSE,
                                      PANGO_ELLIPSIZE_END, FALSE, FALSE, NULL, NULL);
        g_free(text);
      }
      // label on top of marker:
      gchar *label_text = _build_label(w);
      set_color(cr, *text_color);
      const float label_width = inner_w - _widget_get_quad_width(w) - value_width;
      if(label_width > 0)
        show_pango_text(w, context, cr, label_text, 0, 0, label_width, FALSE, FALSE, PANGO_ELLIPSIZE_END, FALSE,
                        TRUE, NULL, NULL);
      g_free(label_text);
    }
    break;
    default:
      break;
  }
```

so good luck finding out what needs a redraw and what doesn't ; logically everything is redrawn all the time, hence the GUI lag.

This needs a rewrite and some heads to fall.

**Comment by jorismak** [comment]
Just a FYI, don't know if you can do something with this info, if it's even related, or that you just think 'f*ck it' 😉 .

Compiled a recent version from your 'candidate' branch, which contains your bauhaus changes.
A lot of things now don't seem to update anymore. I can't 'drag a slider' to see it changing, I have to click it (let go of the mouse button) to see the effect. Also, things like white balance don't seem to update at all or properly anymore.
Also seen a weird effect where (in color calibration module) setting it to 'daylight' gives a different result, depending if you set it while previously on 'A - incandescent' or 'E - equi-energy'.

This is all on Windows.

**Comment by aurelienpierre** [comment]
> A lot of things now don't seem to update anymore. I can't 'drag a slider' to see it changing, I have to click it (let go of the mouse button) to see the effect. Also, things like white balance don't seem to update at all or properly anymore.

Yes, I know. The thing is there will be no fix, I'm in the process of rewriting the whole thing almost entirely. The stack of workarounds made it so fixing something broke the rest, the painting held the walls…

**Comment by pedrorrodriguez** [comment]
Status?

**Comment by aurelienpierre** [comment]
Fixed by 48a86a9c0353f87444fd6956f087af933ecad599
